[#141113589] Postgres event trigger

combor opened the pull request:
## What
This PR introduces event triggers into Postgres to change the ownership of current objects and newly created objects to the dbname_owner role, all new users are added to this role at creation time.

Existing users are migrated to being members of this role when a new user is created during a bind. This builds upon the work created in both #27 by @jmcarp and subsequent work in https://github.com/alphagov/paas-rds-broker/tree/postgres-event-trigger.

Unit tests have been added around this, these are passing in Travis and the blackbox tests are passing on build.

## How to review this

1. Code review
2. Test with the instructions in alphagov/paas-cf#969

## Who can review it
Not me @paroxp, @dcarley or @combor

jmcarp committed:
Manage postgres users with event triggers.

Note: comment from original commit b6c6588c2e406347a10c423b7c06dc65a13e58be
      by @alext and @jonty

Previously the PostgreSQL engine copied the MySQL engine and granted
all rights on the database to the newly created user.

In PostgreSQL this will give the user rights to create tables, but because it
has a more finely-grained permission model than MySQL this does not
give the user rights on existing tables, or tables newly created by other
users. Only the owner of a table can alter/drop it, and you cannot grant this
permission to other users. Users who are the owners of tables cannot
be removed until the table ownership is changed.

Previously we added logic to always return the same user for every binding,
but in this commit we solve all of these problems by creating a "parent"
group that is given ownership of all tables, sequences, and functions.
New users are granted rights on this group and thus can carry out all
owner actions on group-owned items.

Unfortunately there is no way to set a default owner on newly created items (it
defaults to the user that created them), so we install a DDL event trigger
(https://www.postgresql.org/docs/9.5/static/event-trigger-definition.html) that
fires when every `CREATE`, `ALTER`, `DROP`, `GRANT` or `REVOKE` command
finishes. It simply uses the built-in `REASSIGN OWNED BY` command to hand over
ownership of all resources owned by the current user to the parent
`$DBNAME_owner` group.

We use a per-database group name because roles/users/groups are global across
the PostgreSQL instance, and we may wish to support more than one database per
PostgreSQL instance in the future.

We also tried the following approaches:

* Using `ALTER DEFAULT PRIVILEGES`
  (https://www.postgresql.org/docs/9.5/static/sql-alterdefaultprivileges.html) on
  every newly created user to `GRANT ALL ON PUBLIC` - this means that every user
  can `SELECT`, `INSERT`, and `DELETE`, but because only the table owner can
  `ALTER` or `DROP` this will not allow other bound users to run migrations. This
  is also limited to the `PUBLIC` (default) schema, so would fail to work for any
  applications that have custom schemas.

* Making the group the owner of the `PUBLIC` schema. This allowed members of
  the group to `DROP` tables within the schema, but still did not allow them to
  `ALTER` these tables.

jmcarp committed:
Exclude unsupported engine-version combinations.

keymon committed:
reassign_owner() trigger ignores superusers

When running the test against a docker or a travis service, we
create the user with a user with superuser permissions. That user
matches the condition of having access to the group role, which makes
the trigger try to change the ownership of the trigger function itself :)

We need to check if the current user is a superuser, and only reassign
the permissions if it is not.

keymon committed:
ResetState deletes all users

Once we use a dedicated user per binding and a shared group for
ownership, the ResetState() function to apply to restored instances only
needs to delete all the users, except the current user (masteruser) and
any other superuser.

keymon committed:
Update tests to support ownership reassign

Update the sqlengine unit tests to work with the implementation
using triggers that change the object ownership.

We create a new superuser that simulates the master user in RDS, to
simulate the behaviour of RDS and also simulate SQL errors by removing
the permissions of that user.

We add the different cases creating users, dropping them, creating
multiple uses for the same database, etc.

saliceti committed:
User bind/unbind supports legacy bind users

In the previous implementation for postgres we created a unique user
that was returned for every binding, called $dbname+'_owner'.

But in the new implementation we create a user per binding.

In order to support a seamless transition for pre-existing instances
with bindings using the old implementation we add logic that:

 * will not fail on unbinding if the "new style user" does not exist.
   This is because the binding might have been created with the previous
   implementation so the user does not exist.

 * will grant every existing non-superuser user to the new "manager group"
   This group will own all the objects in the new implementation.
   This would allow old bindings to access objects created by new bindings.

 * will reassign all the objects of every existing non-superuser user to
   the new "manager group"
   This would allow the new bindings to access pre-existing objects.

For the transition, we will ask our tenants to rebind the applications.
After a period of grace (e.g. 1 month) we will run a cleanup task to
delete the legacy users and we will revert this code.

henrytk commented on sqlengine/postgres_engine_test.go:
```diff
@@ -168,7 +168,7 @@ func migrationTest(connectionString, dbname string) {
 	}
 
-	usersStatement := "select usename from pg_user where usesuper != true and usename != current_user"
+	usersStatement := "select usename from pg_user where usesuper != true and usename != current_user;"
```
Just out of curiosity: was this breaking something or done for consistency/aesthetic reasons?

dcarley replied:
Discussed with Lee: it was to make it easier to copy/paste the statements into a `psql` shell. While rebasing I've taken the opportunity to remove these because I personally don't think it adds much value, it was making it harder to rebase, and it leaves us with inconsistency between the various statements that do and don't have this. Feel free to discuss with me if you disagree.

henrytk commented on sqlengine/postgres_engine_test.go:
```diff
@@ -414,6 +414,14 @@ var _ = Describe("PostgresEngine", func() {
 				connectionStringNew = postgresEngine.URI(address, port, dbname, createdUser, createdPassword)
 			})
 
+			AfterEach(func() {
```
This whole commit looks like it should be squashed into whatever introduced the tests this is running after each of.

dcarley replied:
I think this has been done now (though I'm lacking some context on the original comment).

henrytk reviewed (commented):
I don't think it is clear enough what commits can be reverted once we no longer need the migration code. It looks like they could be squashed together and marked as removable once the right conditions have been met. Organising the code into revertable commits will be much easier now when it is fresh in our minds and will reduce the chance of things being missed later.

henrytk commented on .travis.yml:
```diff
@@ -8,7 +8,7 @@ services:
 
 addons:
-  postgresql: "9.3"
+  postgresql: "9.4"
```
Three things:

1. we use Postgres 9.5 for our RDS broker created instances. Is there a reason for not going with 9.5?
2. there is another commit which bumps the Postgres version in Travis in this pull request, they can probably be squashed.
3. the commit message is "Update Travis config to accomodate new tests. " - what new tests? I don't understand what that means.

dcarley replied:
I've clarified this in a separate commit 6ff6d176123544751cd507299da2a57edef41224

henrytk commented on sqlengine/postgres_engine.go:
```diff
@@ -1,43 +1,55 @@
 )
 
-const defaultStateDBName = "broker_state"
+var ensureTriggerPattern = `
```
I'm not sure why these template patterns were made as globals. This may just be my opinion, but I found this really interrupted the readability of the package code. Also I can't see a reason why they need to be global, since they are not referenced anywhere else in the repository.

LeePorte replied:
Refactored into respective functions

alext replied:
The advantage of doing this in the global scope is that the templates only get compiled once at startup instead of being compiled every time.

dcarley replied:
There's still global now, but they've moved closer to the functions that use them. I think that's fine for the reason Alex describes.

dcarley committed:
Update Travis config to use Postgres 9.5

We want the tests run by Travis CI to use the same version of Postgres that:

- supports event triggers, as used by the modified bind/unbind behaviour in
  preceding commits
- matches the version used in our catalogs for integration tests and
  production

According to the output of previous builds it was defaulting to 9.1:

    $ sudo service postgresql start
     * Starting PostgreSQL 9.1 database server

9.5 is not available on Container Precise, which is the default:

- https://docs.travis-ci.com/user/database-setup/#Using-a-different-PostgreSQL-Version

So we also need to switch to using Container Trusty:

- https://docs.travis-ci.com/user/trusty-ci-environment/#Using-Trusty

dcarley committed:
Fix typo in encryption plan change error.

dcarley committed:
Refactoring postgres_engine for event triggers

Collection of minor changes that didn't warrant separate commits but should
be separate from the migration code:

- move trigger statement variables closer to their usage so that you don't
  need to jump around the file to read them. They are still global vars
  because it means the templates only need to be rendered once, but the
  string can be a constant
- separate listing non-superusers to a private function so that it's easier
  to read and can be re-used by migration code
- split apart generateUsername() to multiple function calls to make it
  easier to read
